File: tools/pnnx/tests/test_storezip.py
#!/usr/bin/env python3

# Copyright 2026 Tencent
# SPDX-License-Identifier: BSD-3-Clause

import io
import struct
import subprocess
import sys
import tempfile
import unittest
import zipfile
from pathlib import Path


STOREZIP_HELPER = Path(sys.argv[1])
sys.argv = [sys.argv[0]]


class NonSeekableBuffer(io.BytesIO):
    def seek(self, *args, **kwargs):
        raise io.UnsupportedOperation("seek")


def run_helper(*args):
    return subprocess.run(
        [str(STOREZIP_HELPER), *map(str, args)],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        check=False,
    )


def write_zip(path, entries, compression=zipfile.ZIP_STORED):
    with zipfile.ZipFile(path, "w", compression=compression) as archive:
        for name, data in entries.items():
            archive.writestr(name, data)


def write_data_descriptor_zip(path, entries):
    stream = NonSeekableBuffer()
    with zipfile.ZipFile(stream, "w", compression=zipfile.ZIP_STORED) as archive:
        for name, data in entries.items():
            archive.writestr(name, data)
    path.write_bytes(stream.getvalue())


class StoreZipReaderTest(unittest.TestCase):
    def test_reads_stored_entry_with_data_descriptor(self):
        with tempfile.TemporaryDirectory() as temp_dir:
            archive_path = Path(temp_dir) / "model.pt2"
            entry_name = "arbitrary-root/archive_format"
            write_data_descriptor_zip(archive_path, {entry_name: b"pt2"})

            with zipfile.ZipFile(archive_path) as archive:
                self.assertNotEqual(archive.getinfo(entry_name).flag_bits & 0x08, 0)

            result = run_helper("read", archive_path, entry_name)
            self.assertEqual(result.returncode, 0, result.stderr.decode(errors="replace"))
            self.assertEqual(result.stdout, b"pt2")

    def test_reopen_discards_entries_from_previous_archive(self):
        with tempfile.TemporaryDirectory() as temp_dir:
            first_path = Path(temp_dir) / "first.zip"
            second_path = Path(temp_dir) / "second.zip"
            write_zip(first_path, {"old-entry": b"old"})
            write_zip(second_path, {"new-entry": b"new"})

            result = run_helper(
                "reopen-without-stale-entry",
                first_path,
                second_path,
                "old-entry",
            )
            self.assertEqual(result.returncode, 0, result.stderr.decode(errors="replace"))

    def test_reads_storezip_writer_archive(self):
        with tempfile.TemporaryDirectory() as temp_dir:
            archive_path = Path(temp_dir) / "writer.zip"
            result = run_helper("writer-round-trip", archive_path)
            self.assertEqual(result.returncode, 0, result.stderr.decode(errors="replace"))

    def test_rejects_archive_without_eocd(self):
        with tempfile.TemporaryDirectory() as temp_dir:
            archive_path = Path(temp_dir) / "missing-eocd.zip"
            write_zip(archive_path, {"payload": b"data"})
            data = archive_path.read_bytes()
            eocd_offset = data.rfind(b"PK\x05\x06")
            self.assertGreaterEqual(eocd_offset, 0)
            archive_path.write_bytes(data[:eocd_offset])

            result = run_helper("expect-open-failure", archive_path)
            self.assertEqual(result.returncode, 0, result.stderr.decode(errors="replace"))

    def test_rejects_central_directory_outside_file(self):
        with tempfile.TemporaryDirectory() as temp_dir:
            archive_path = Path(temp_dir) / "invalid-offset.zip"
            write_zip(archive_path, {"payload": b"data"})
            data = bytearray(archive_path.read_bytes())
            eocd_offset = data.rfind(b"PK\x05\x06")
            self.assertGreaterEqual(eocd_offset, 0)
            struct.pack_into("<I", data, eocd_offset + 16, len(data) + 4096)
            archive_path.write_bytes(data)

            result = run_helper("expect-open-failure", archive_path)
            self.assertEqual(result.returncode, 0, result.stderr.decode(errors="replace"))

    def test_rejects_compressed_entry(self):
        with tempfile.TemporaryDirectory() as temp_dir:
            archive_path = Path(temp_dir) / "deflated.zip"
            write_zip(archive_path, {"payload": b"data"}, zipfile.ZIP_DEFLATED)

            result = run_helper("expect-open-failure", archive_path)
            self.assertEqual(result.returncode, 0, result.stderr.decode(errors="replace"))


if __name__ == "__main__":
    unittest.main()

File: tools/pnnx/tests/test_storezip.cpp
// Copyright 2026 Tencent
// SPDX-License-Identifier: BSD-3-Clause

#include "storezip.h"

#include <stdio.h>
#include <string.h>

#include <algorithm>
#include <string>
#include <vector>

static int read_entry(const char* path, const char* name)
{
    pnnx::StoreZipReader reader;
    if (reader.open(path) != 0)
        return 1;

    if (!reader.has_file(name))
        return 1;

    const uint64_t size = reader.get_file_size(name);
    std::vector<char> data(size);
    if (reader.read_file(name, data.data()) != 0)
        return 1;

    if (!data.empty())
        fwrite(data.data(), 1, data.size(), stdout);

    return 0;
}

static int expect_open_failure(const char* path)
{
    pnnx::StoreZipReader reader;
    return reader.open(path) == 0 ? 1 : 0;
}

static int expect_read_failure_after_open(const char* path, const char* name)
{
    pnnx::StoreZipReader reader;
    if (reader.open(path) != 0)
        return 1;

    if (!reader.has_file(name))
        return 1;

    const uint64_t size = reader.get_file_size(name);
    std::vector<char> data(size);
    return reader.read_file(name, data.data()) == 0 ? 1 : 0;
}

static int expect_compressed_entry(const char* path, const char* name)
{
    pnnx::StoreZipReader reader;
    if (reader.open(path) != 0)
        return 1;

    if (!reader.has_file(name))
        return 1;

    return reader.is_file_stored(name) ? 1 : 0;
}

static int reopen_without_stale_entry(const char* first_path, const char* second_path, const char* stale_name)
{
    pnnx::StoreZipReader reader;
    if (reader.open(first_path) != 0)
        return 1;
    if (reader.open(second_path) != 0)
        return 1;

    const std::vector<std::string> names = reader.get_names();
    if (std::find(names.begin(), names.end(), stale_name) != names.end())
        return 1;

    return reader.has_file(stale_name) ? 1 : 0;
}

static int writer_round_trip(const char* path)
{
    static const char payload[] = "storezip-round-trip";

    pnnx::StoreZipWriter writer;
    if (writer.open(path) != 0)
        return 1;
    if (writer.write_file("payload", payload, sizeof(payload) - 1) != 0)
        return 1;
    if (writer.close() != 0)
        return 1;

    pnnx::StoreZipReader reader;
    if (reader.open(path) != 0)
        return 1;
    if (!reader.has_file("payload"))
        return 1;
    if (!reader.is_file_stored("payload"))
        return 1;
    if (reader.get_file_size("payload") != sizeof(payload) - 1)
        return 1;

    std::vector<char> actual(sizeof(payload) - 1);
    if (reader.read_file("payload", actual.data()) != 0)
        return 1;

    return memcmp(actual.data(), payload, actual.size()) == 0 ? 0 : 1;
}

int main(int argc, char** argv)
{
    if (argc < 3)
        return 2;

    const std::string command = argv[1];

    if (command == "read" && argc == 4)
        return read_entry(argv[2], argv[3]);
    if (command == "expect-open-failure" && argc == 3)
        return expect_open_failure(argv[2]);
    if (command == "expect-read-failure-after-open" && argc == 4)
        return expect_read_failure_after_open(argv[2], argv[3]);
    if (command == "expect-compressed-entry" && argc == 4)
        return expect_compressed_entry(argv[2], argv[3]);
    if (command == "reopen-without-stale-entry" && argc == 5)
        return reopen_without_stale_entry(argv[2], argv[3], argv[4]);
    if (command == "writer-round-trip" && argc == 3)
        return writer_round_trip(argv[2]);

    return 2;
}
